pkg/adhocprofiles/adhocprofiles.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	bucket, err := a.getBucket(tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	uid := ulid.MustNew(ulid.Timestamp(adHocProfile.UploadedAt), rand.Reader)
+	id := strings.Join([]string{uid.String(), adHocProfile.Name}, "-")
+
+	dataToStore, err := json.Marshal(adHocProfile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to upload profile")
+	}
+
+	err = bucket.Upload(ctx, id, bytes.NewReader(dataToStore))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to upload profile")
+	}
+
+	jsonProfile, err := json.Marshal(profile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	return connect.NewResponse(&v1.AdHocProfilesGetResponse{
+		Id:                 id,
+		Name:               adHocProfile.Name,
+		UploadedAt:         adHocProfile.UploadedAt.UnixMilli(),
+		FlamebearerProfile: string(jsonProfile),
+		ProfileType:        profile.Metadata.Name,
+		ProfileTypes:       profileTypes,
+	}), nil
+}
+
+func (a *AdHocProfiles) Get(ctx context.Context, c *connect.Request[v1.AdHocProfilesGetRequest]) (*connect.Response[v1.AdHocProfilesGetResponse], error) {
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	bucket, err := a.getBucket(tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := bucket.Get(ctx, c.Msg.GetId())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get profile")
+	}
+
+	adHocProfileBytes, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var adHocProfile AdHocProfile
+	err = json.Unmarshal(adHocProfileBytes, &adHocProfile)
+	if err != nil {
+		return nil, err
+	}
+
+	maxNodes, err := validation.ValidateMaxNodes(a.limits, []string{tenantID}, c.Msg.GetMaxNodes())
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not determine max nodes")
+	}
+
+	profile, profileTypes, err := parse(&adHocProfile, c.Msg.ProfileType, maxNodes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	jsonProfile, err := json.Marshal(profile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	return connect.NewResponse(&v1.AdHocProfilesGetResponse{
+		Id:                 c.Msg.Id,
+		Name:               adHocProfile.Name,
+		UploadedAt:         adHocProfile.UploadedAt.UnixMilli(),
+		FlamebearerProfile: string(jsonProfile),
+		ProfileType:        profile.Metadata.Name,
+		ProfileTypes:       profileTypes,
+	}), nil
+}
+
+func (a *AdHocProfiles) List(ctx context.Context, c *connect.Request[v1.AdHocProfilesListRequest]) (*connect.Response[v1.AdHocProfilesListResponse], error) {
+	bucket, err := a.getBucketFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	profiles := make([]*v1.AdHocProfilesProfileMetadata, 0)
+	err = bucket.Iter(ctx, "", func(s string) error {
+		separatorIndex := bytes.IndexByte([]byte(s), '-')
```

Comment:
```suggestion
		separatorIndex := strings.IndexRune(s, '-')
```

This is to avoid []byte conversion which copies the string, it's irrelevant here but I just wanted to let you know if you ever do that in a hot path be aware of string <-> []byte conversion.